modules/core/include/opencv2/core/quaternion.inl.hpp

```diff
@@ -0,0 +1,851 @@
+    {
+        S = std::sqrt(1.0 - R(0, 0) + R(1, 1) - R(2, 2)) * 2;
+        x = (R(0, 1) + R(1, 0)) / S;
+        y = 0.25 * S;
+        z = (R(1, 2) + R(2, 1)) / S;
+        w = (R(0, 2) - R(2, 0)) / S;
+    }
+    else
+    {
+        S = std::sqrt(1.0 - R(0, 0) - R(1, 1) + R(2, 2)) * 2;
+        x = (R(0, 2) + R(2, 0)) / S;
+        y = (R(1, 2) + R(2, 1)) / S;
+        z = 0.25 * S;
+        w = -(R(0, 1) - R(1, 0)) / S;
+    }
+    return Quat<T> (w, x, y, z);
+}
+
+template <typename T>
+Quat<T> Quat<T>::createFromRvec(InputArray _rvec)
+{
+    if (!((_rvec.cols() == 1 && _rvec.rows() == 3) || (_rvec.cols() == 3 && _rvec.rows() == 1))) {
+        CV_Error(Error::StsBadArg, "Cannot convert rotation vector to quaternion: The length of rotation vector should be 3");
+    }
+    Vec<T, 3> rvec;
+    _rvec.copyTo(rvec);
+    T psi = std::sqrt(rvec.dot(rvec));
+    if (abs(psi) < CV_QUAT_EPS) {
+        return Quat<T> (1, 0, 0, 0);
+    }
+    Vec<T, 3> axis = rvec / psi;
+    return createFromAngleAxis(psi, axis);
+}
+
+template <typename T>
+inline Quat<T> Quat<T>::operator-() const
+{
+    return Quat<T>(-w, -x, -y, -z);
+}
+
+
+template <typename T>
+inline bool Quat<T>::operator==(const Quat<T> &q) const
+{
+    return (abs(w - q.w) < CV_QUAT_EPS && abs(x - q.x) < CV_QUAT_EPS && abs(y - q.y) < CV_QUAT_EPS && abs(z - q.z) < CV_QUAT_EPS);
+}
+
+template <typename T>
+inline Quat<T> Quat<T>::operator+(const Quat<T> &q1) const
+{
+    return Quat<T>(w + q1.w, x + q1.x, y + q1.y, z + q1.z);
+}
+
+template <typename T>
+inline Quat<T> Quat<T>::operator-(const Quat<T> &q1) const
+{
+    return Quat<T>(w - q1.w, x - q1.x, y - q1.y, z - q1.z);
+}
+
+template <typename T>
+inline Quat<T>& Quat<T>::operator+=(const Quat<T> &q1)
+{
+    w += q1.w;
+    x += q1.x;
+    y += q1.y;
+    z += q1.z;
+    return *this;
+}
+
+template <typename T>
+inline Quat<T>& Quat<T>::operator-=(const Quat<T> &q1)
+{
+    w -= q1.w;
+    x -= q1.x;
+    y -= q1.y;
+    z -= q1.z;
+    return *this;
+}
+
+template <typename T>
+inline Quat<T> Quat<T>::operator*(const Quat<T> &q1) const
+{
+    // FIXIT !!! '*' and '*=' must provide similar results to avoid users confusions
+    Vec<T, 4> q{w, x, y, z};
+    Vec<T, 4> q2{q1.w, q1.x, q1.y, q1.z};
+    return Quat<T>(q * q2);
+    // return Quat<T>(w * q1.w, x * q1.x, y * q1.y, z * q1.z);
+}
+
+
+template <typename T, typename S>
+Quat<T> operator*(const Quat<T> &q1, const S a)
+{
+    return Quat<T>(a * q1.w, a * q1.x, a * q1.y, a * q1.z);
+}
+
+template <typename T, typename S>
+Quat<T> operator*(const S a, const Quat<T> &q1)
+{
+    return Quat<T>(a * q1.w, a * q1.x, a * q1.y, a * q1.z);
+}
+
+template <typename T>
+inline Quat<T>& Quat<T>::operator*=(const Quat<T> &q1)
+{
+    T qw, qx, qy, qz;
+    qw = w * q1.w - x * q1.x - y * q1.y - z * q1.z;
+    qx = x * q1.w + w * q1.x + y * q1.z - z * q1.y;
+    qy = y * q1.w + w * q1.y + z * q1.x - x * q1.z;
+    qz = z * q1.w + w * q1.z + x * q1.y - y * q1.x;
```

Comment:
In general users expect the same operations under `operator*()` and `operator*=()`
```
c = a * b;
a *= b;
ASSERT(a == c);
```

This is not true for current implementation.

If this can't be unified, then it is better to undefine these operators and use dedicated well-named and documented functions.

Reply:
I'm so sorry I didn't repeat the mistake you mentioned. Here is my test code:
```
Quatd a{-2.1, 1.7, 0, 3.1415926};
Quatd b{2.5, 0, -2, 0.5};
Quatd c = a * b;
a *= b;
CV_Assert(a == c);
```
But I found no issue here.
Could you please describe more details?

Reply:
You are right. My mistake (forgot about cross product for `Vec` type).